### ISSUE TITLE
Unable to stream non MP4 files.

### ISSUE THREAD

**Issue by OstlerDev** [OstlerDev/PopcornTV#3]
Because of a limitation with the Apple TV it is unable to stream anything but MP4 files, meaning that any AVI, MKV, or alternate videos need to be converted. The most supported way will require use of FFMPEG and will need to be run inline with the torrent downloader.

**Comment by OstlerDev** [comment]
Added error XML to display when file is not of mp4 type. 359f29c5faf94e213e74123ec911697c12d97a61

**Comment by OstlerDev** [comment]
There are a few ways that I have been thinking of doing this, here they are for future reference. Please note that I may have tested a bit but have not gotten anything working as of yet.

- Using the ffmpeg segementor, segment the file into lots of little .ts files, then send a .m3u8 file to the Apple TV as it converts. Resume the segmentor at the last .ts before error to continue encoding.
- Convert the actual source file that is fullsized (?), this will throw an error when it attempts to read somewhere that has not been written. Meaning that we need a way to track the frame written and pause the encoder, then resume it. (control + z on mac pauses the process)
- Hope that we are downloading fast enough and start the stream after we have at least 20mb converted.

**Comment by OstlerDev** [comment]
We can convert the video as it is being downloaded by capping the loadrate to the frame rate. Meaning that it will take one second to convert one second of actual video, this introduces a problem if the user fastforwards, as it is not able to convert ahead of the user.

```
ffmpeg -re -i input.mkv -movflags +faststart output.mp4
```

**Comment by OstlerDev** [comment]
The commit 6cf2e9c implements basic support for converting files, there are still glaring issues like below:

- It currently shows the current time instead of Time remaining.
- If it is not downloading fast enough then it will only partially convert (cannot really be helped).

This needs lots of testing. If you wish to test it, please install FFMPEG: http://ffmpeg.org/download.html

**Comment by JakeRound** [comment]
Very nice work Ostler! Unfortunately i was unable to get it running, I expected this being a very buggy being it's in early development  it's nice to see you're able to start working on the torrent conversion system. Only if you had a PayPal i'd gladly donate to support the project.

**Comment by OstlerDev** [comment]
@JakeRound Since it is still early in development, attaching a bug log can only help :)

Also, you can buy Bitcoin with Paypal, it is just a little pricier. I cannot accept Paypal for a couple reasons, but the main one being that they closed my account permanently.

**Comment by masonjames** [comment]
I'd be happy to help with debugging this too.

I haven't used ffmpeg since like... 2007.  I'm on OS X.  What's the best way for me to test?
I see recently updated libraries here: http://www.osxexperts.net/ffmpeg/ffmpegexperts.html
And gui here: http://www.iffmpeg.com/FFmpegInfo/index.html

I think I used to use this previously: http://www.ffmpegx.com/download.html

What's the best way to test/report? Is there a particular tv episode, for example, to use as a good sample?  Or some indicator within PopcornTV where I can check that it isn't an mp4?

Lemme know. Excited to see such quick iterations on this one :)

**Comment by OstlerDev** [comment]
I added an article on the Wiki on how to install FFMPEG, try to follow that, if the instructions are not clear enough let me know.

Most TV shows that are 720p or 1080p are not in MP4 formats.

PopcornTV will output something when it attempts to convert the file. If you run it in Debug mode you will see a lot of stuff output as it converts.

**Comment by masonjames** [comment]
So I followed the OS X FFMPEG procedures, and here's the debug file from a fresh attempt:
https://gist.github.com/masonjames/1cbcb44d38f67076cef6

Key takeaway: 'Error: Cannot find module \'parse-torrent\''

**Comment by OstlerDev** [comment]
Did you make sure to run "npm install"?

**Comment by masonjames** [comment]
I did.  Tried a couple times with the same result.

**Comment by OstlerDev** [comment]
Make sure to pull down all the changes, the module is inside the [package.json file](https://github.com/OstlerDev/PopcornTV/blob/master/package.json#L27).

If that does not work, just run ``npm install parse-torrent``

**Comment by masonjames** [comment]
My God. It works!

I finally ran `sudo npm install parse-torrent` because I was getting an access error running it normally.

I then selected 2 different tv shows. In each case I chose the 720p versions which had not worked previously and in both cases they played extremely well.  I am running PopcornTV with debug so if the logs are helpful (though lengthy) I'll share them.

I will say I'm running a macbook pro with a 2.9 GHz Intel Core i5 and the CPU usage ran at about 200% consistently for the ffmpeg process.  Memory usage was fine clocking in at 200MB for ffmpeg and 150 for node.

The wait time between clicking 'play' and the episode beginning to stream is about twice with conversion as it is when it just plays the regular mp4.  That's extremely subjective and makes sense. I can see where users who aren't watching the logs might think something was up, so just noting that here.

It also seems like it downloads/converts as much as possible - as fast as it possibly can.  This provides the most-likely uninterrupted viewing process, which is awesome. You might consider a setting for throttling it a bit for folks who don't want to max out their CPUs - or that might be entirely on ffmpeg to do so.  Again, just making notes - something like that would clearly be "having your cake and eating it too"

Crazy good stuff here. What else can I do to help?

**Comment by OstlerDev** [comment]
@masonjames I have figured out a way to better post what is happening when opening a stream and I am going to implement that soon I hope (If I can get it to work reliably). As for the CPU maxing out, there is no way that I can slow down the encode further, right now it only encodes one second of video per second. In order to stop the stream from ending, this is how it has to be.

**Comment by OstlerDev** [comment]
There is an issue where it crashes after starting the conversion, further looking into needed: https://gist.github.com/OstlerDev/72b5a06287868184ffbd

**Comment by rimkashox** [comment]
@OstlerDev just so you know... There is a system that works well, using peerflix to download the torrent and drag drop to beamer to play to apple tv. This combo works well to my knowledge. I have done it countless times. When the download it slow, the wheel turns a little then it resumes.. Too bad beamer is closed source. 

How to try:
...
peerflix [TORRENT FILE/MAGNET LINK] --remove
...
Open Finder and go to /tmp/torrent-stream/<path to file found in peerflix output>
Drag file into Beamer and go make some popcorn to watch your video with.

**Comment by OstlerDev** [comment]
Peerflix just downloads the torrent and passes it into a viewer, we have something similar already and so we don't need the extra bulk.

Beamer uses an FFMPEG encoder, I tried decompiling to see how they did it, but I was unsuccessful in my attempt.

**Comment by rimkashox** [comment]
Glad you tried that already..  Peerflix does not pass to a viewer per say, it write the torrent file to disk pure and simple, but beamer is able to handle this file even if it's not complete and loading... It seems that it reads from the file incrementally, maybe frame by frame ? (don't know much about this) Also maybe we can try sniffing the traffic between beamer and apple tv, to see if there's something special in the way it streams...

**Comment by OstlerDev** [comment]
@rimkashox The way I believe they are doing it is directly passing the torrent data into the converter using a pipe, where as we are loading the file which can throw an error when it tries to read a chunk that has not been written to. We have looked into methods to directly pass the torrent data stream into the converter bypassing writing it into a file, but it does not work as well.

We will still be trying, but for now I suggest people use 480p until this gets sorted out.

**Comment by rimkashox** [comment]
@OstlerDev With the peerflix + beamer system... peeflix writes let's say to rambo-720p.mkv directly... After letting it load for 10 seconds, I drag and drop this file "not yet complete" file into beamer... and beamer doesn't know that rambo-720p.mkv is not complete, so that's kind of a hack.. But it works ! Maybe some versions of ffmpeg can handle incomplete files.. I think that's the way, not cutting the file into many smaller ones like the .ts you are using now..

**Comment by OstlerDev** [comment]
Peerflix actually uses torrent-stream (which we use), I believe beamer might do something similar to how I currently do it, but more testing would be needed. They have error handling. Right now I just set the program to crash when there is an error (as this feature is in the alpha stage).

I have talked with ffmpeg dev's and they all said similar things that the only way to combat this would be to slow it down to only converting one second of video every second.

**Comment by rimkashox** [comment]
Yes !! Second by second or even frame by frame if doable.. but the output shouldn't be .ts files in my opinion.. but a real stream.. That's my intuition. You are on the right track !

**Comment by OstlerDev** [comment]
Well we are doing it second for second already, as slow as we can. The output is .ts files because that way we can actually stream it to the TV. I might end up being forced to write a whole other project just to handle things like this.

**Comment by rimkashox** [comment]
Yes, if you do then this piece of code can solve the Apple TV popcorn time problem too... This would be such a great contribution.
